code/crates/core-types/src/signing.rs

```diff
@@ -201,48 +201,135 @@ where
-            let Some(validator) = validator_set.get_by_address(&commit_sig.address) else {
-                return Err(CertificateError::UnknownValidator(commit_sig.clone()));
-            };
+            let validator = validator_set
+                .get_by_address(validator_address)
+                .ok_or_else(|| CertificateError::UnknownValidator(validator_address.clone()))?;
 
-            let voting_power = self.verify_commit_signature(certificate, commit_sig, validator)?;
-            signed_voting_power += voting_power;
+            if let Ok(voting_power) =
+                self.verify_commit_signature(certificate, commit_sig, validator)
+            {
+                signed_voting_power += voting_power;
+            }
         }
 
+        let total_voting_power = validator_set.total_voting_power();
+
+        // Check if we have 2/3+ voting power
+        if thresholds
+            .quorum
+            .is_met(signed_voting_power, total_voting_power)
+        {
+            Ok(())
+        } else {
+            Err(CertificateError::NotEnoughVotingPower {
+                signed: signed_voting_power,
+                total: total_voting_power,
+                expected: thresholds.quorum.min_expected(total_voting_power),
+            })
+        }
+    }
+
+    /// Verify the polka certificate against the given validator set.
+    ///
+    /// - For each signed prevote in the certificate:
+    ///   - Reconstruct the signed prevote and verify its signature
+    /// - Check that we have 2/3+ of voting power has signed the certificate
+    ///
+    /// If any of those steps fail, return a [`CertificateError`].
+    ///
+    fn verify_polka_certificate(
+        &self,
+        certificate: &PolkaCertificate<Ctx>,
+        validator_set: &Ctx::ValidatorSet,
+        thresholds: ThresholdParams,
+    ) -> Result<(), CertificateError<Ctx>> {
+        use crate::ValidatorSet;
+
+        let mut signed_voting_power = 0;
+        let mut seen_validators = Vec::new();
+
+        // For each vote, check that the validator has not already vote and verify the signature
+        for vote in &certificate.votes {
+            let validator_address = vote.validator_address();
+
+            if seen_validators.contains(&validator_address) {
+                return Err(CertificateError::DuplicateVote(validator_address.clone()));
+            }
+
+            seen_validators.push(validator_address);
+
+            // Abort if validator not in validator set
+            let validator = validator_set
+                .get_by_address(validator_address)
+                .ok_or_else(|| CertificateError::UnknownValidator(validator_address.clone()))?;
+
+            let valid_vote =
+                self.verify_signed_vote(&vote.message, &vote.signature, validator.public_key());
+
+            // Only tally valid votes for a value that matches the certificate
+            if valid_vote && vote.value().as_ref() == NilOrVal::Val(&certificate.value_id) {
```

Comment:
We should also check the vote height and round match the certificate's. Ideally we should "aggregate" like we do for commit certificate.

Reply:
Added the check in 0d64776b100059660afe8d7066273b7c91f9cd64.

Will do the aggregation as an another commit.

Reply:
503aa421411e9a27116b53ed1578cc9761aaeb72